content/blog/2025-05-03-blog.md

```diff
@@ -0,0 +1,51 @@
+Consequently, modern-day browsers have to use JIT compilers to get maximal performance gains.
+Since JavaScript is a dynamically-typed language, a traditional compiler must emit code that can handle the different possible types during runtime, producing less efficient code overall; JIT compilers are not bound by this limitation.
+But how can you make an effective JIT for JavaScript?
+
+The authors of the paper expect that programs spend most of their time in hot loops, and that hot loops are generally type-stable — this is their key idea.
+In the paper, they describe the development of a JIT compiler built around this expectation called TraceMonkey.
+
+At a high level, TraceMonkey works by first monitoring for hot loops by using counters.
+Once hot loops are identified, it checks for type stability throughout the loop.
+If it is type stable, then it can emit corresponding code that is closer to efficiency to that of a statically-typed language.
+It can also detect violations of this assumption and resolve them at runtime.
+
+---
+
+TraceMonkey typically outperformed the original SpiderMonkey interpreter on their provided benchmarks, but the improvements were largely inconsistent. Benchmarks focused on bitwise operations had nearly 25x speedups; others showed no improvements. TraceMonkey performed especially badly on programs with a lot of recursion because only loops were tracked.
+
+---
+
+TraceMonkey is only one of several different JIT implementations used in SpiderMonkey, the name of Mozilla’s JavaScript engine used to power Firefox.
+Although TraceMonkey was great at tracing loops and generating efficient code, it (for obvious or aforementioned reasons) was weak in processing programs that did not have as many loops or used a lot of recursion.
+In particular, TraceMonkey had no method-based compiling and could suffer from trace explosion.
+For these and other reasons, TraceMonkey has long been deprecated, but it directly influenced its successors: JagerMonkey, WarpMonkey, and IonMonkey. 
+Each of these compilers use more method-based JIT compiling while still incorporating trace-based compilation in a hybrid approach. Similarly, JavaScript JIT compilers for other popular browsers — Chrome and Safari — also used a method-based approach. Is this to say trace-based JITs are less effective than method-based JITs? Not quite; the ideas presented in TraceMonkey still live on in several ways. In addition to the previously stated fact that the newer Firefox JITs still use tracing, the predominant JIT compilers for Python and Lua (PyPy and LuaJIT respectively) all heavily rely on tracing, with LuaJIT considered being one of the fastest JITs ever. 
+It may be helpful to go into more detail about the transition to focusing on method tracing. 
+Trace-based compiling excels when control flow is simpler, but on the web, control flow can be highly irregular for several reasons: dynamic input, user interaction, etc. 
```

Comment:
> Trace-based compiling excels when control flow is simpler

I'm not sure I agree with this assertion, or at least I think you haven't presented enough evidence for the assertion to be convincing. There is a lot of branching but those branches are "stable," then it seems like tracing is a wonderful way to automatically eliminate a lot of it in the common case.